oxide-auth/src/endpoint/tests/client_credentials.rs

```diff
@@ -0,0 +1,250 @@
+        urlbody: Some(
+            vec![("grant_type", "client_credentials")]
+                .iter()
+                .to_single_value_query(),
+        ),
+        auth: Some(format!("Basic {}", setup.basic_authorization)),
+    };
+
+    setup.test_success(success, Allow(EXAMPLE_CLIENT_ID.to_owned()));
+}
+
+#[test]
+fn client_credentials_success_changed_owner() {
+    let mut setup = ClientCredentialsSetup::new();
+    let success = CraftedRequest {
+        query: None,
+        urlbody: Some(
+            vec![("grant_type", "client_credentials")]
+                .iter()
+                .to_single_value_query(),
+        ),
+        auth: Some(format!("Basic {}", setup.basic_authorization)),
+    };
+
+    setup.test_success(success, Allow("OtherOwnerId".to_owned()));
+}
+
+#[test]
+fn client_credentials_deny_public_client() {
+    let mut setup = ClientCredentialsSetup::public_client();
+    let public_client = CraftedRequest {
+        query: None,
+        urlbody: Some(
+            vec![
+                ("grant_type", "client_credentials"),
+                ("client_id", EXAMPLE_CLIENT_ID),
+            ]
+            .iter()
+            .to_single_value_query(),
+        ),
+        auth: None,
+    };
+
+    setup.test_bad_request(public_client, Deny);
+}
+
+#[test]
+fn client_credentials_deny_incorrect_credentials() {
+    let mut setup = ClientCredentialsSetup::new();
+    let basic_authorization = base64::encode(&format!("{}:the wrong passphrase", EXAMPLE_CLIENT_ID));
+    let wrong_credentials = CraftedRequest {
+        query: None,
+        urlbody: Some(
+            vec![("grant_type", "client_credentials")]
+                .iter()
+                .to_single_value_query(),
+        ),
+        auth: Some(format!("Basic {}", basic_authorization)),
+    };
+
+    setup.test_unauthorized(wrong_credentials, Allow(EXAMPLE_CLIENT_ID.to_owned()));
+}
+
+#[test]
+fn client_credentials_deny_missing_credentials() {
+    let mut setup = ClientCredentialsSetup::new();
+    let missing_credentials = CraftedRequest {
+        query: None,
+        urlbody: Some(
+            vec![
+                ("grant_type", "client_credentials"),
+                ("client_id", EXAMPLE_CLIENT_ID),
+            ]
+            .iter()
+            .to_single_value_query(),
+        ),
+        auth: None,
+    };
+
+    setup.test_bad_request(missing_credentials, Allow(EXAMPLE_CLIENT_ID.to_owned()));
+}
+
+#[test]
+fn client_credentials_deny_unknown_client() {
+    // The client_id is not registered
+    let unknown_client = CraftedRequest {
+        query: None,
+        urlbody: Some(
+            vec![
+                ("grant_type", "client_credentials"),
+                ("client_id", "SomeOtherClient"),
+            ]
+            .iter()
+            .to_single_value_query(),
+        ),
+        auth: None,
+    };
+
+    ClientCredentialsSetup::new().test_bad_request(unknown_client, Allow("SomeOtherClient".to_owned()));
+}
```

Comment:
It relates to the indistinguishabiltiy shortly touched upon before. During an authentication attempt it shouldn't be possible to distinguish between an unknown user and a bad password. Otherwise, an attacker might perform user enumeration by trying usernames.

This request, if done properly, should fail the same way as a bad password does (with `unauthenticated`). It currently doesn't because it's not actually sending a password—which is an invalid request. That property can be computed by the attacker so it's not a leak to respond this way. There should be a second test with a valid but unauthenticated other client id.